S3 upload and download

opadron opened the pull request:
This extends Spack functionality so that it can fetch sources and binaries from-, push sources and binaries to-, and index the contents of- mirrors hosted on an S3 bucket.

Rough to-do list:

- [x] Extend mirrors configuration to add support for `file://`, and `s3://` URLs.
- [x] Ensure all fetching, pushing, and indexing operations work for `file://` URLs.
- [x] Implement S3 source fetching
- [x] Implement S3 binary mirror indexing
- [x] Implement S3 binary package fetching
- [x] Implement S3 source pushing
- [x] Implement S3 binary package pushing

opadron commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -1063,6 +1064,23 @@ def for_package_version(pkg, version):
     raise InvalidArgsError(pkg, version, **args)
 
 
+def from_url_scheme(url, *args, **kwargs):
```
I noticed that the original code assumed all mirrors were `http(s)://` URLs, and it occurred to me that the actual fetch strategy that should be used would depend on the URLs `scheme`.  It wasn't clear to me where I might find a function for constructing a fetch strategy based on a URL's scheme, so I added one, here.

opadron commented on lib/spack/spack/schema/mirrors.py:
```diff
@@ -17,7 +17,19 @@
         'additionalProperties': False,
         'patternProperties': {
-            r'\w[\w-]*': {'type': 'string'},
+            r'\w[\w-]*': {
```
Mirrors can now have separate `push` and `fetch` URLs (like git), for the case where a mirror may be accessed via a mechanism that differs from that with which resources are pushed to it, such as a mirror accessed through a web server, but populated by copying files to a path on the local file system.

opadron commented on lib/spack/spack/stage.py:
```diff
@@ -381,8 +381,13 @@ def fetch(self, mirror_only=False):
                         url, digest, expand=expand, extension=extension))
+                # fetchers.insert(
+                #     0, fs.URLFetchStrategy(
+                #         url, digest, expand=expand, extension=extension))
```
Here, the URL `scheme`-based function is used to determine what fetch strategy to employ.

Question: In testing, I noticed that if one the strategies throws, the whole fetch operation fails.  I understand that what is preferred is that every URL should be tried (with their associated fetch strategies), and only if they *all* fail to download the source should the fetch operation fail.  What do others think?  If this behavior is, indeed, preferred, where would be the best places to make the necessary changes?

opadron replied:
I think this change, as-is, is the likely cause of #13404.

My intent was, put simply, to allow fetching from `s3://` URLs by providing `S3FetchStrategy`.  The issue is that the original logic here, where the fetch strategy is selected, was problematic because it effectively always used `URLFetchStrategy` before trying anything else, and appending the S3 strategy did not work because, as I noted above, the whole fetch operation would fail when the first attempt using `URLFetchStrategy` raised an exception.

What I ended up doing was just checking the URLs `scheme` and matching it against a known table of fetch strategies.  This change may have led to the issue found in #13404, though I'm not yet sure how.

opadron replied:
I thought a little more about the changes, and I think fetching web URLs are now going through urllib (with these changes).  I believe before these changes spack was running `curl` to perform the downloads.  I'm not 100% sure that my recollection is correct, but if it is, that difference is very likely to be the specific change that introduced #13404.

@scheibelp

opadron commented on lib/spack/spack/util/url.py:
```diff
@@ -0,0 +1,166 @@
+
+"""
+Utility functions for parsing, formatting, and manipulating URLs.
+"""
```
I ended up doing a lot of custom URL parsing/manipulation.  I also noticed other places in the code that would be simpler with utilities for these operations.  So, I added a few in this new module.

opadron commented on lib/spack/spack/s3_handler.py:
```diff
@@ -0,0 +1,180 @@
+        tuple(''.join((A, 'ccept', sep, R, 'anges'))
+            for A, sep, R in
+            it.product('Aa', [''] + list(' _-'), 'Rr')),
+
```
Note to self: remove duplicate.

opadron commented on lib/spack/spack/util/web.py:
```diff
@@ -243,13 +264,63 @@ def _spider_wrapper(args):
 
 
-def _urlopen(*args, **kwargs):
+# TODO(opadron): There's gotta be a better place for stuff like this.
```
@tgamblin @scheibelp @alalazo @adamjstewart @scottwittenburg Thoughts?

(Also, is there a way to mention reviewers, like `@reviewers`?  Would be nice.)

tgamblin replied:
I don't understand what the purpose of `noopify` is or why we need it.  Can you explain what you're really trying to do here?

opadron replied:
I was writing an explanation, but in doing so I decided that it was just way too complicated for what we need.  I think I should just get rid of it.

tgamblin commented on lib/spack/spack/s3_fetch_strategy.py:
```diff
@@ -0,0 +1,91 @@
+    @_needs_stage
+    def fetch(self):
+        # delay import to avoid circular imports
+        from spack.util.web import read_from_url
```
don't do this.  Just don't use `from`.

tgamblin commented on lib/spack/spack/s3_fetch_strategy.py:
```diff
@@ -0,0 +1,91 @@
+                   "you are trying to use does not exist or you have an "
+                   "internet gateway issue. You can remove the bad archive "
+                   "using 'spack clean <package>', then try again using "
+                   "the correct URL.")
```
Don't just copy/paste code like this.  If the s3 fetcher needs this message (I think it probably does?), then factor it out and refer to it from both the URL fetcher and s3 fetcher, but don't make 2 copies that we have to maintain twice.

tgamblin commented on lib/spack/spack/s3_handler.py:
```diff
@@ -0,0 +1,173 @@
+from llnl.util.lang import memoized
+
+import spack
+from spack.util.url import parse as urlparse, join as urljoin
```
just `import spack.util.url as url` and prefer to use the namespace.

tgamblin commented on lib/spack/spack/s3_handler.py:
```diff
@@ -0,0 +1,173 @@
+        return len(self.__set())
+
+
+HTTP_HEADERS_KEY_ALIASES = (
```
This needs a comment or an explanation or something.  It's not at all clear why it needs to be so ugly.

opadron replied:
In testing, I found that the response objects returned by boto for S3 objects, urllib for "file://" URLs, and urllib for other URLs all had variations in how headers were spelled (e.g.: "ContentType" vs "Content-Type" vs "content_type"), which was throwing key errors for calling sites that tried to use a different spelling.  The mapping with aliases class is my attempt at wrapping the headers in a dict for which `d['ContentType']` would just work regardless of how the underlying dict spelled it.

tgamblin commented on lib/spack/spack/util/web.py:
```diff
@@ -86,25 +98,39 @@ def __init__(self, *args, **kwargs):
 
-def _read_from_url(url, accept_content_type=None):
+__UNABLE_TO_VERIFY_SSL = (
+        lambda pyver: (
```
why is there a `lambda` in here?  Why not just make this a clearly written function?

tgamblin commented on lib/spack/spack/util/s3.py:
```diff
@@ -0,0 +1,51 @@
+
+    # if no access credentials provided above, then access anonymously
+    if not session.get_credentials():
+        from botocore import UNSIGNED
```
Avoid nested imports.  `import botocore as bc`

tgamblin commented on lib/spack/spack/util/url.py:
```diff
@@ -0,0 +1,230 @@
+from spack.util.path import canonicalize_path
+
+
+class ParseResultWrapper(ParseResult):
```
docstring

tgamblin commented on lib/spack/spack/util/url.py:
```diff
@@ -0,0 +1,230 @@
+            ' ({})'.format(str(self.__dict__)) if self.__dict__ else ''))
+
+
+def _split_all(path):
```
docstring (also can this be made simpler?)

tgamblin commented on lib/spack/spack/util/url.py:
```diff
@@ -0,0 +1,230 @@
+
+
+def parse(url, scheme='file'):
+    """Parse a mirror url."""
```
Better docstring -- document return type.  Also document why we need this instead of just urlparse.

tgamblin commented on lib/spack/spack/util/url.py:
```diff
@@ -0,0 +1,230 @@
+    return result
+
+
+def format(parsed_url):
```
docstring

tgamblin commented on lib/spack/spack/util/url.py:
```diff
@@ -0,0 +1,230 @@
+    #
+    # ... they would be (correctly) formatted like so:
+    #
+    #  s3:///my-bucket
```
Are s3 URLs with 3 slashes a thing?  I was skimming this and didn't see any:

https://gpdb.docs.pivotal.io/43320/admin_guide/load/topics/g-s3-protocol.html

Also does Boto not provide any utilities to deal with some of the things this function is doing?

opadron replied:
> @opadron: I haven't had a chance to go over all the questions here, but I added a bunch of review comments. In general, the high level plan looks good, and the refactors are good. However:
> 
> 1. A lot of the mirror code is repetitive and overly complex. I recommend introducing a `Mirror` class to take care of that.  Currently too much of the code deals directly with YAML instead of with a `Mirror` abstraction.

Agreed.

> 2. Boxes are checked on the PR, but I don't see where the commands now have push or pull options.  Where is that implemented?

Fetching and pushing is implemented directly in the mirror and buildcache commands.  With these changes, `s3://` URLs (or mirror names referring to such URLs) can be provided anywhere where these commands expected a directory.  In these cases, the commands work exactly as they did before, except they handle uploading to- and/or downloading from- S3 buckets in a way that is transparent to the user.

For example: `spack buildcache create -d s3:///spack-public/mirror zlib` populates (creating, if necessary) a buildcache with zlib binaries.  It works exactly the same as before, except now it also works with S3 buckets.

> 3. We discussed making `/var/spack/cache` the "default" mirror where build caches would be created and where things would be created if no `--directory` argument is specified for `spack mirror create` and `spack buildcache create`.  Is that planned?
> 4. There is a lot of stuff in here for parsing URLs and for things like `noopify`, and I am not sure how it factors in or why some of it is needed.  Maybe we can talk about that at an upcoming meeting.

opadron replied:
> Are s3 URLs with 3 slashes a thing? I was skimming this and didn't see any:
> 
> https://gpdb.docs.pivotal.io/43320/admin_guide/load/topics/g-s3-protocol.html
> 
> Also does Boto not provide any utilities to deal with some of the things this function is doing?

To my knowledge there isn't a standard defining what an S3 URL should have or how it should be formatted.  I tried defining a format that could encode all the information we might need to make the connection with what I hoped are reasonable defaults.  The format documented in your link is actually very similar to what I've come up with.  I will update with documentation for the format, which should answer this and possibly other questions you might have about it.

Boto does not provide any handling of S3 URLs because they're not officially a thing.

tgamblin replied:
If they're not officially a thing, then we shouldn't invent a new thing.  It looks like S3 URLs are pretty recognizable as they are now, e.g., [from here](https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingBucket.html) it looks like there are three forms:
* http://<bucket>.s3-<aws-region>.amazonaws.com
* http://<bucket>.s3.amazonaws.com
* http://s3-<aws-region>.amazonaws.com/<bucket>

If those are already copy/pastable from somewhere in AWS, I do not think we need a new URL scheme.  We can infer what we need to know about the bucket from the https URL.

opadron replied:
The "Accessing a Bucket" section in the docs you link fails to cover a lot of important details. 
 They don't really clarify what is meant by "access", only having this relevant piece to say:

> If you access a bucket programmatically, note that Amazon S3 supports RESTful architecture in which your buckets and objects are resources, each with a resource URI that uniquely identifies the resource.

It says nothing about what actual operations are supported or what headers need to be included for authentication, access control, or billing purposes.  This omission also appears deliberate, as all the AWS documentation I've read on the matter prefers to point you to their console UI, CLI, SDKs or boto.

Regardless, even if this documentation did exist (which it must, somewhere), we'd still need to provide extra information with each request (likely through headers) that is not encoded in any of the URL examples you point to, nor in any examples I've seen in the wild, other than those like the ones in the pivotal docs you linked above.

There's no way an http URL, on its own, is ever going to have enough information for Spack to work with the buckets they may refer to.  And the story gets worse once you consider non-AWS endpoints.

I'm not a fan of inventing a new URL scheme, either, but there's only so much we can do to avoid it if we want to work with S3 buckets.  My hope is that we can take some comfort in that

1. There is some prior art (just nothing standardized or in library form) that I used as-is as my starting point, so I'm not really "inventing" all that much, just reimplementing, and
1. The code is already written and works.  The URL scheme just needs to be documented.

tgamblin commented on lib/spack/spack/util/web.py:
```diff
@@ -36,7 +43,12 @@ class HTMLParseError(Exception):
+from spack.util.s3 import create_s3_session
+from spack.util.url import join as urljoin, parse as urlparse
+
+from spack.s3_handler import open as s3_open
```
We should rethink the module breakdown here, as there are a lot of circular imports.  In general we want to avoid any circular dependencies between module in Spack (I know there are already quite a few)

tgamblin commented on lib/spack/spack/util/web.py:
```diff
@@ -390,7 +607,7 @@ def get_checksums_for_versions(
                 ["{0:{1}}  {2}".format(str(v), max_len, url_dict[v])
                  for v in sorted_versions]))
-    print()
+    tty.msg('')
```
why?  print() adds an empty line, while `tty.msg()` prints an arrow.

opadron replied:
That's a think-o on my part.  Will fix.

opadron commented:
@tgamblin I've rebased and added two new commits.  Adding the `Mirror` and `MirrorCollection` classes went a long way towards cleaning up a lot of the more minor points you had brought up.  Have a look and let me know if you think I'm on the right track.

eugeneswalker commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -23,14 +24,32 @@
+</html>
+'''
+
+BUILD_CACHE_INDEX_ENTRY_TEMPLATE = '  <li><a href="{path}">{path}</a>'
```
Shouldn't we close out the `<li>` tag here?

opadron replied:
Good catch!

paulbry commented on lib/spack/spack/util/web.py:
```diff
@@ -113,29 +175,184 @@ def _read_from_url(url, accept_content_type=None):
+
+    verify_ssl = spack.config.get('config:verify_ssl')
+
+    if __UNABLE_TO_VERIFY_SSL and verify_ssl and uses_ssl(remote_path):
```
```suggestion
    if __UNABLE_TO_VERIFY_SSL and verify_ssl and uses_ssl(remote_url):
```

Believe this should be changed, currently experiencing a `Warning: 'str' object has no attribute 'scheme'`.

opadron replied:
Thanks for that, @paulbry!

scottwittenburg commented on lib/spack/spack/util/s3.py:
```diff
@@ -0,0 +1,50 @@
+
+
+def create_s3_session(url):
+    parsed_url = urllib_parse.urlparse(
```
I'm hitting an issue here when I attempt to use `util/web.py` to `push_to_url()` a file.  Below is a some output illustrating the problem, but basically, it seems like sometimes the `url` argument here may have already been parsed as a url.

```
scott@beast:/data/scott/projects/spack$ export AWS_ACCESS_KEY_ID=minio
scott@beast:/data/scott/projects/spack$ export AWS_SECRET_ACCESS_KEY=minio123
scott@beast:/data/scott/projects/spack$ export AWS_ENDPOINT_URL=http://localhost:8083
scott@beast:/data/scott/projects/spack$ source share/spack/setup-env.sh 
scott@beast:/data/scott/projects/spack$ spack python
Spack version 0.12.1
Python 3.6.8, Linux x86_64
>>> import spack.util.web as web_util
>>> dir(web_util)
['ALLOWED_ARCHIVE_TYPES', 'HTMLParseError', 'HTMLParser', 'HTTP_HEADER_NAME_ALIASES', 'LinkParser', 'NoNetworkConnectionError', 'NonDaemonContext', 'NonDaemonPool', 'NonDaemonProcess', 'Request', 'SpackWebError', 'URLError', '__UNABLE_TO_VERIFY_SSL', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', '_iter_s3_prefix', '_list_s3_objects', '_separators', '_spider', '_spider_wrapper', '_timeout', '_urlopen', 'codecs', 'errno', 'find_versions_of_archive', 'list_url', 'mkdirp', 'multiprocessing', 'os', 'print_function', 'product', 'push_to_url', 're', 'read_from_url', 'remove_url', 's3_util', 'shutil', 'spack', 'spider', 'ssl', 'standardize_header_names', 'string_types', 'sys', 'traceback', 'tty', 'url_exists', 'url_util', 'urllib_parse', 'urlopen', 'uses_ssl', 'warn_no_ssl_cert_checking']
>>> local_url = 'file:///tmp/junk.txt'
>>> remote_url = 's3://junkbox/stuff/junk.txt'
>>> web_util.push_to_url(local_url, remote_url)
Traceback (most recent call last):
  File "/usr/lib/python3.6/code.py", line 91, in runcode
    exec(code, self.locals)
  File "<console>", line 1, in <module>
  File "/data/scott/projects/spack/lib/spack/spack/util/web.py", line 254, in push_to_url
    s3 = s3_util.create_s3_session(remote_url)
  File "/data/scott/projects/spack/lib/spack/spack/util/s3.py", line 17, in create_s3_session
    allow_fragments=False)
  File "/usr/lib/python3.6/urllib/parse.py", line 367, in urlparse
    url, scheme, _coerce_result = _coerce_args(url, scheme)
  File "/usr/lib/python3.6/urllib/parse.py", line 120, in _coerce_args
    raise TypeError("Cannot mix str and non-str arguments")
TypeError: Cannot mix str and non-str arguments
>>> 
now exiting InteractiveConsole...
scott@beast:/data/scott/projects/spack$ spack python
Spack version 0.12.1
Python 3.6.8, Linux x86_64
>>> import spack.util.web as web_util
>>> local_url = 'file:///tmp/junk.txt'
>>> remote_url = 's3://junkbox/stuff/junk.txt'
>>> web_util.push_to_url(local_url, remote_url)
> /data/scott/projects/spack/lib/spack/spack/util/s3.py(16)create_s3_session()
-> parsed_url = urllib_parse.urlparse(
(Pdb) l
 11  	
 12  	
 13  	def create_s3_session(url):
 14  	    import pdb
 15  	    pdb.set_trace()
 16  ->	    parsed_url = urllib_parse.urlparse(
 17  	        url,
 18  	        scheme='file',
 19  	        allow_fragments=False)
 20  	
 21  	    if parsed_url.scheme != 's3':
(Pdb) p url
ParseResult(scheme='s3', netloc='junkbox', path='/stuff/junk.txt', params='', query='', fragment='')
(Pdb)
```

scottwittenburg replied:
The following patch fixed it for me and I was able to push the local file into the S3 bucket:

```
scott@beast:/data/scott/projects/spack$ git diff
diff --git a/lib/spack/spack/util/s3.py b/lib/spack/spack/util/s3.py
index 2d0a34a29..095b057b9 100644
--- a/lib/spack/spack/util/s3.py
+++ b/lib/spack/spack/util/s3.py
@@ -11,10 +11,12 @@ import spack
 
 
 def create_s3_session(url):
-    parsed_url = urllib_parse.urlparse(
-        url,
-        scheme='file',
-        allow_fragments=False)
+    parsed_url = url
+    if not isinstance(parsed_url, urllib_parse.ParseResult):
+        parsed_url = urllib_parse.urlparse(
+            url,
+            scheme='file',
+            allow_fragments=False)
 
     if parsed_url.scheme != 's3':
         raise ValueError(
scott@beast:/data/scott/projects/spack$
```

scottwittenburg replied:
To be clear, I'm not suggesting that is necessarily the right solution, but hopefully it makes clear what could be going wrong.

tgamblin commented:
@opadron: this is still failing CI.

tgamblin commented:
@scheibelp: I think this is going to have a few conflicts with #12940 -- can you take a look at this one and make a recommendation about which we should merge first?  I *think* it may be easier to put this in first, then adapt #12940 around it, but I could be wrong.

scottwittenburg commented:
@opadron @tgamblin I just wanted to chime in here that after the latest commits, this branch is working for me.  I'm using it mainly in the context of build pipelines where it's working for pushing buildcaches to mirrors, installing dependencies from them, and downloading buildcache files from the remote binary mirror to the local artifacts buildcache.

FYI @aashish24

scheibelp commented:
> I think it may be easier to put this in first, then adapt #12940 around it

That works for me. My only thought on which one to merge first is that merging the bigger one first might save some time (and this one is bigger). That being said if review for this takes a while I wouldn't consider it problematic to merge #12940 first (but there are still review comments to address there).

aashish24 commented:
thanks @scottwittenburg @opadron what do you think? If all looks good, I would think it is better to merge this as this one is long standing @scheibelp and may help with release

opadron commented:
Yeah, sorry if I wasn't clear, but in the absence of any further requests for changes, this should be ready for merge.

scheibelp commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -449,7 +460,10 @@ def archive(self, destination):
         if not self.archive_file:
             raise NoArchiveFileError("Cannot call archive() before fetching.")
 
-        shutil.copyfile(self.archive_file, destination)
```
@opadron I think the following is causing #13404:

* The archive file is determined in part by the basename of the URL, in this case `libuuid-1.0.3.tar.gz?r=http%3A%2F%2Fsourceforge.net%2Fprojects%2Flibuuid%2F&ts=1433881396&use_mirror=iweb`
* Originally the `shutil.copyfile` command happily used this as the source name of the file (which was weird, but it worked)
* Now the `web.push_to_url` function is used, and it modifies the supplied local path with `url_util.parse(local_path)`

I think that

* `web.push_to_url` should not format the input path (the local file is what it is) i.e. remove the `url_util.parse(local_path)`
* arguably the archive filename should be formatted better (but technically that is not required) which is to say that the archive name should be based on `url_util.parse` of the basename of the URL

Would just removing `url_util.parse(local_path)` cause problems that you know of?
